File: Sources/UIView/UIView+Snapshot.m
//
//  UIView+Snapshot.m
//  SPUIKitCategory
//
//  Created by LSP on 15/12/10.
//  Copyright © 2015年 LSP. All rights reserved.
//

#import "UIView+Snapshot.h"

@implementation UIView (Snapshot)
- (UIImage *)snapImage {
    return [self.layer snapImage];
}
- (UIImage *)snapImage:(BOOL)opaque {
    return [self.layer snapImage:opaque];
}
@end


@implementation CALayer (Snapshot)

- (UIImage *)snapImage {
    return [self snapImage:NO];
}

- (UIImage *)snapImage:(BOOL)opaque {
    
    if(/* DISABLES CODE */ (&UIGraphicsBeginImageContextWithOptions) != NULL)
    {
        UIGraphicsBeginImageContextWithOptions(self.frame.size, opaque, 0.0);
    } else {
        UIGraphicsBeginImageContext(self.frame.size);
    }
    
    [self renderInContext:UIGraphicsGetCurrentContext()];
    UIImage *img = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    return img;
}

@end

File: Sources/UIView/UIView+Snapshot.swift
//
//  UIView+Snapshot.swift
//  SPUIKitCategory module
//
//  Created by LSP on 2020/12/15.
//  Copyright © 2020 LSP. All rights reserved.
//

#if !os(macOS)
import UIKit

extension CALayer {
    public
    func snapImage(opaque: Bool = false) -> UIImage {
        UIGraphicsBeginImageContextWithOptions(self.frame.size, opaque, 0.0)
        if let ctx = UIGraphicsGetCurrentContext() {
            self.render(in: ctx)
            if let img = UIGraphicsGetImageFromCurrentImageContext() {
                UIGraphicsEndImageContext()
                return img
            }
        }
        UIGraphicsEndImageContext()
        return UIImage()
    }
}

extension UIView {
    public
    func snapImage(opaque: Bool = false) -> UIImage {
        return self.layer.snapImage(opaque: opaque)
    }
}
#endif
